.github/workflows/test_functional.yml

```diff
@@ -106,54 +133,79 @@ jobs:
           ssh -vv _remote_background_indep_poll hostname
 
       - name: Test
+        id: test
+        timeout-minutes: 35
+        continue-on-error: true
         run: |
-          # Use non-UTC time zone
-          export TZ=XXX-05:30
-
-          if [[ "${BASE}" == 'tests/k' ]]; then
+          if [[ '${{ matrix.test-base }}' == 'tests/k' ]]; then
               NPROC=4
           else
               NPROC=8
           fi
-          # NOTE: $BASE is purposefully un-quoted
+          # NOTE: test base is purposefully un-quoted
           etc/bin/run-functional-tests \
               -j "${NPROC}" \
-              --state=save ${BASE} \
-          || (
-              echo '=== Re-running Failed Tests ==='
-              CHUNK= etc/bin/run-functional-tests \
-                  -j "${NPROC}" \
-                  --state=save,failed
-          )
+              --state=save \
+              ${{ matrix.test-base }} \
+              || (touch tests-did-not-time-out && false)
+
+      - name: Re-run failed tests
+        timeout-minutes: 10
+        if: steps.test.outcome == 'failure'
+        run: |
+          if [[ ! -f tests-did-not-time-out ]]; then
+            echo "the preious test run timed out -> not re-running tests"
+            # this false causes the workflow to fail
+            false
+          fi
+          # TODO: make the tests deterministic so we don't need to do this
+          etc/bin/run-functional-tests \
+              -j 1 \
+              -v \
+              --state=save,failed ${{ matrix.test-base }}
```

Comment:
This re-runs any failed tests with `-j 1 -v` which will give us much more useful information for debugging.

I had resisted doing this before as it makes it possible for flaky tests that are not parallel-safe to pass as they get run in series the second time around. However, it's not like the way we were doing it before was much better from that perspective. Just need to keep an eye on flaky tests and look to improve/replace them where possible.